File: packages/python-sdk/src/urisocial/resources/custom_guides.py
"""Custom visual guides resource for URI Social SDK"""

from typing import Dict, Any, List, Optional
from ..http_client import HTTPClient


class CustomGuidesResource:
    """Custom visual guide management - reference image analysis and font matching"""

    def __init__(self, http: HTTPClient):
        self._http = http

    def upload_reference_image(
        self,
        image_url: str,
        name: str,
        brand_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Upload and process reference image to create custom visual guide

        Performs:
        - Safety, quality, copyright screening
        - Aesthetic extraction (GPT-4o-mini Vision)
        - Typography extraction and font matching
        - Prompt assembly
        - 11-dimension metadata tagging

        Args:
            image_url: Cloudinary URL of uploaded reference image
            name: Guide name (1-100 chars)
            brand_id: Optional brand ID

        Returns:
            Guide preview with match outcome for user confirmation
        """
        return self._http.post(
            "/social-media/custom-guides/upload",
            json={
                "image_url": image_url,
                "name": name,
                "brand_id": brand_id,
            },
        )

    def list_guides(
        self,
        brand_id: Optional[str] = None,
        status: Optional[str] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> Dict[str, Any]:
        """
        List custom visual guides

        Args:
            brand_id: Optional brand ID to filter by
            status: Optional status filter ('active', 'archived')
            limit: Max number of results (default: 50)
            offset: Pagination offset (default: 0)

        Returns:
            List of guide objects
        """
        params = {"limit": limit, "offset": offset}
        if brand_id:
            params["brand_id"] = brand_id
        if status:
            params["status"] = status
        return self._http.get("/social-media/custom-guides", params=params)

    def get_guide(self, guide_id: str) -> Dict[str, Any]:
        """
        Get specific custom visual guide

        Args:
            guide_id: Guide ID

        Returns:
            Guide object with full details
        """
        return self._http.get(f"/social-media/custom-guides/{guide_id}")

    def update_guide_font(self, guide_id: str, matched_font_id: str) -> Dict[str, Any]:
        """
        Update guide's matched font

        Args:
            guide_id: Guide ID
            matched_font_id: New font ID to use

        Returns:
            Updated guide object
        """
        return self._http.patch(
            f"/social-media/custom-guides/{guide_id}/font",
            json={"matched_font_id": matched_font_id},
        )

    def delete_guide(self, guide_id: str) -> Dict[str, Any]:
        """
        Delete (archive) custom visual guide

        Args:
            guide_id: Guide ID

        Returns:
            Deletion confirmation
        """
        return self._http.delete(f"/social-media/custom-guides/{guide_id}")

    def rematch_typography(self, guide_id: str) -> Dict[str, Any]:
        """
        Re-run typography matching for guide

        Useful when font library is updated or initial match was poor.

        Args:
            guide_id: Guide ID

        Returns:
            Updated guide object with new typography match
        """
        return self._http.post(f"/social-media/custom-guides/{guide_id}/rematch")

    def auto_rematch_all(
        self,
        brand_id: Optional[str] = None,
        min_confidence_threshold: float = 0.5,
    ) -> Dict[str, Any]:
        """
        Automatically re-match typography for all guides below confidence threshold

        Args:
            brand_id: Optional brand ID to filter by
            min_confidence_threshold: Only rematch guides below this confidence (0.0-1.0, default: 0.5)

        Returns:
            Summary of rematch operation
        """
        params = {"min_confidence_threshold": min_confidence_threshold}
        if brand_id:
            params["brand_id"] = brand_id
        return self._http.post("/social-media/custom-guides/auto-rematch", params=params)

    def track_usage(self, guide_id: str) -> Dict[str, Any]:
        """
        Track usage of custom visual guide

        Increments times_used counter for analytics.

        Args:
            guide_id: Guide ID

        Returns:
            Usage confirmation
        """
        return self._http.post(f"/social-media/custom-guides/{guide_id}/track-usage")

    def cleanup_archived(self, days_threshold: int = 90) -> Dict[str, Any]:
        """
        Cleanup archived guides older than threshold

        Args:
            days_threshold: Delete archived guides older than this many days (default: 90)

        Returns:
            Cleanup summary
        """
        return self._http.post(
            "/social-media/custom-guides/cleanup-archived",
            params={"days_threshold": days_threshold},
        )

File: packages/typescript-sdk/src/resources/custom-guides.ts
import { HTTPClient } from '../client/http';

// ── Types ──────────────────────────────────────────────────────────────

export interface CustomGuide {
  id: string;
  user_id: string;
  brand_id?: string;
  name: string;
  original_image_url: string;
  uploaded_at: string;
  aesthetic_summary: Record<string, any>;
  typography_match: Record<string, any>;
  match_outcome: string;
  metadata_tags: Record<string, any>;
  times_used: number;
  status: 'active' | 'archived';
  created_at: string;
  updated_at: string;
}

export interface UploadReferenceImageRequest {
  image_url: string;
  name: string;
  brand_id?: string;
}

export interface UpdateGuideFontRequest {
  matched_font_id: string;
}

export interface ListGuidesParams {
  brand_id?: string;
  status?: 'active' | 'archived';
  limit?: number;
  offset?: number;
}

export interface AutoRematchParams {
  brand_id?: string;
  min_confidence_threshold?: number;
}

export interface CleanupArchivedParams {
  days_threshold?: number;
}

// ── Resource ───────────────────────────────────────────────────────────

export class CustomGuidesResource {
  constructor(private http: HTTPClient) {}

  /**
   * Upload and process reference image to create custom visual guide
   *
   * Performs:
   * - Safety, quality, copyright screening
   * - Aesthetic extraction (GPT-4o-mini Vision)
   * - Typography extraction and font matching
   * - Prompt assembly
   * - 11-dimension metadata tagging
   */
  async uploadReferenceImage(request: UploadReferenceImageRequest): Promise<{ guide: CustomGuide }> {
    return this.http.post<{ guide: CustomGuide }>('/social-media/custom-guides/upload', request);
  }

  /**
   * List custom visual guides
   */
  async listGuides(params?: ListGuidesParams): Promise<{ guides: CustomGuide[] }> {
    return this.http.get<{ guides: CustomGuide[] }>('/social-media/custom-guides', { params });
  }

  /**
   * Get specific custom visual guide
   */
  async getGuide(guideId: string): Promise<{ guide: CustomGuide }> {
    return this.http.get<{ guide: CustomGuide }>(`/social-media/custom-guides/${guideId}`);
  }

  /**
   * Update guide's matched font
   */
  async updateGuideFont(guideId: string, request: UpdateGuideFontRequest): Promise<{ guide: CustomGuide }> {
    return this.http.patch<{ guide: CustomGuide }>(`/social-media/custom-guides/${guideId}/font`, request);
  }

  /**
   * Delete (archive) custom visual guide
   */
  async deleteGuide(guideId: string): Promise<{ success: boolean }> {
    return this.http.delete<{ success: boolean }>(`/social-media/custom-guides/${guideId}`);
  }

  /**
   * Re-run typography matching for guide
   *
   * Useful when font library is updated or initial match was poor.
   */
  async rematchTypography(guideId: string): Promise<{ guide: CustomGuide }> {
    return this.http.post<{ guide: CustomGuide }>(`/social-media/custom-guides/${guideId}/rematch`);
  }

  /**
   * Automatically re-match typography for all guides below confidence threshold
   */
  async autoRematchAll(params?: AutoRematchParams): Promise<{ summary: Record<string, any> }> {
    return this.http.post<{ summary: Record<string, any> }>('/social-media/custom-guides/auto-rematch', null, {
      params,
    });
  }

  /**
   * Track usage of custom visual guide
   *
   * Increments times_used counter for analytics.
   */
  async trackUsage(guideId: string): Promise<{ success: boolean }> {
    return this.http.post<{ success: boolean }>(`/social-media/custom-guides/${guideId}/track-usage`);
  }

  /**
   * Cleanup archived guides older than threshold
   */
  async cleanupArchived(params?: CleanupArchivedParams): Promise<{ summary: Record<string, any> }> {
    return this.http.post<{ summary: Record<string, any> }>('/social-media/custom-guides/cleanup-archived', null, {
      params,
    });
  }
}
